### DISCUSSION TITLE
CPU usage meets an limitaion during Milvus Query

### DISCUSSION THREAD

**Discussion by jiangkai1109** [milvus-io/milvus#25571] [Q&A and General discussion]
HW: 240 vcpu server
SW: milvus 2.2.11 Standalone
Benchmark: vectordb-benchmark  ./benchmark.sh -d gist-960-euclidean -e milvus -c 100 -s 127.0.0.1 -i

case 1- milvus on 240 vCPU: With the above benchmark, 80 vCPU will be occupied by Milvus query. When modify concurrency=200 in the benchmark config file, it still uses 80 vCPU. No CPU usage increasing. 
case 2 - milvus binding on 120 vCPU: With the above benchmark, 50 vCPU will be occupied by Milvus query.  Using the same workload, but the CPU for query is not 80vCPU. 

Expectation: CPU usage will be increased, when concurrency gets higher.
Issue:           CPU usage will not increase when make concurrency larger than 100. For case 2, CPU number working for query is lower(50vcpu) than the case1 cpu number(80vcpu). Suppose they should be the same with the same benchmarking.

Query: Any parameter limit the CPU usage? How can I make 240vCPU full utiliazation during query?

Thanks
Kai

**Comment by yhmo** [comment]
The vectordb-benchmark repo is out of maintenance, maybe is not compatible with the latest milvus version. We have built a new benchmark tool here: https://github.com/zilliztech/VectorDBBench
It is written in Python, easy to read and test. You can use this tool to test your cases.

**Reply by jiangkai1109** [reply]
Thanks for quick reply. Acctually, per the test results, the old vectordb-benchmark works well for the test. Performance scaling well with 1, 2, 4, 8 threads and cores. (I met some issues to use the new benchmark to connect with my local milvus server, and the dataset is a bit small for the test)

While, what I observed is milvus 2.2.11 Standalone can not use more cpu cores: 80 vCPU used out of 240 cores, 50vCPU used out of 120 vCPU. It seems there's some parameters make a limitation of total cpu usage.
Could you share any parameters for the milvus server configuration, which may impact the total CPU usage for the query case?

Thanks
Kai

**Reply by yhmo** [reply]
Some items that might affect CPU usage:
```
queryNode:
  scheduler:
    cpuRatio: 10.0 # ratio used to estimate read task cpu usage.
  grouping:
    enabled: true
    maxNQ: 50000
    topKMergeRatio: 10.0
```
Inside the search engine, a search request CPU usage is estimated by cpuRatio * nq * segment_count. If the estimated usage is small, the search engine will allow multiple search requests to be executed parallelly. 

Assume a query node has 16 cores, we say the CPU ability is 1600. The collection has 5 segments. Now you send 100 search requests at the same time with nq=8. cpuRatio=10. The search engine estimates usage of each request to be 8*5*10=400. Then it allows 1600/400=4 requests to be executed at the same time.

maxNQ is for search requests combination. The search engine can combine similar search requests into a big request to execute. For example, there are 100 requests(nq=5, topk=10) received at the same time. The search engine merges the 100 requests into one request with nq=500, topk=10 to execute. A big request is easy to get higher CPU usage.

You can change the value of cpuRatio and maxNQ to observe CPU usage.

A similar question is here: https://github.com/milvus-io/milvus/discussions/25550

**Reply by jiangkai1109** [reply]
Thanks for suggestions. I found these parameters in file milvus/configs/milvus.yaml.
May I know how can I input this configuration file milvus.yaml during milvus starting? And how to confirm new parameters are imported for the milvus booting up?(by dump?)

I used below steps to start milvus standalone:

Running Milvus standalone
$ cd deployments/docker/dev
$ docker-compose up -d
$ cd ../../../
$ ./scripts/start_standalone.sh

**Reply by jiangkai1109** [reply]
Made below parameters modification for milvus docker, while, the total cpu usage keeps at ~50%.

For milvus docker, modified below parameters in milvus.yaml
cacheSize: 256
maxReadConcurrentRatio: 20
maxNQ: 500000

And add below lines in docker-compose.yml to let docker use milvus.yaml configuration.
  standalone:
    volumes:
      - /root/jk/milvus/milvus_docker/milvus.yaml:/milvus/configs/milvus.yaml

**Reply by yhmo** [reply]
If you start local milvus with " ./scripts/start_standalone.sh", it uses the config file under the source code project milvus/configs/milvus.yaml.  Just modify this file before you run the script  ./scripts/start_standalone.sh

change the value of cpuRatio and maxNQ to observe CPU usage.

**Reply by jiangkai1109** [reply]
I modified the below paramers both for docker and dev mode, but no obvious cpu usage improvement.
maxReadConcurrentRatio: 20 //or 50
maxNQ: 500000

Also tried below cases, for 16 vCPU case, the cpu usage can reach 83%. While, for 112 vCPU case, the cpu usage is only 46%. I am wondering any other paramers impact the cpu usage, such as memory parameter limitation or others? (current system memory is 1TB)
If binding milvus to 16vCPU, the max cpu usage can reach 83%(13vCPU)
If binding milvus to 112vCPU(1 socket), the max cpu usage is limited to 46%(52vCPU).

Parameters confirmation method:
For the docker mode,  connecting to docker via command "docker exec -it milvus_standalone", then checking the parameters in file /milvus/configs/milvus.yaml. These parameters are modified as expected.

For dev mode, not sure how to check the parameters importing or not.

**Comment by xiaofan-luan** [comment]
> I modified the below paramers both for docker and dev mode, but no obvious cpu usage improvement. maxReadConcurrentRatio: 20 //or 50 maxNQ: 500000
> 
> Also tried below cases, for 16 vCPU case, the cpu usage can reach 83%. While, for 112 vCPU case, the cpu usage is only 46%. I am wondering any other paramers impact the cpu usage, such as memory parameter limitation or others? (current system memory is 1TB) If binding milvus to 16vCPU, the max cpu usage can reach 83%(13vCPU) If binding milvus to 112vCPU(1 socket), the max cpu usage is limited to 46%(52vCPU).
> 
> Parameters confirmation method: For the docker mode, connecting to docker via command "docker exec -it milvus_standalone", then checking the parameters in file /milvus/configs/milvus.yaml. These parameters are modified as expected.
> 
> For dev mode, not sure how to check the parameters importing or not.

Hi Jiangkai,
1. we never test milvus on large machines more than 64 core pod
2. we do say the over 16 core, cpu can not be fully utilized.
3. I think the main reason is the request scheduler not fully utilize all cpu resources, we are still investigate on that,if you can offer more details (like pprof and perf) it would be really helpful.
4. be aware of the NUMA issue. it would be better if we split 240 core into several pod.

Feel free to connect me at xiaofan.luan@zilliz.com so we can have more details.

**Reply by jiangkai1109** [reply]
There are 2 numa node of the CPU. 
For the case above, I bind milvus with socket 0 CPUs and membind=0, which equals binding with numa node 0. 
Also did test for 2 sockets acrossing numa0/1. The workload are balanced acrossing all the cores, but the total cpu usage is limited below 50%. I will send the perf info to your mail for discussion. Thanks.
